### ISSUE TITLE
Lab 4: Failing tc 6 output 

### ISSUE THREAD

**Issue by jeffkhong11** [nus-cs2030/2223-s2#177]
# Summary

Description of the question / problem here.

## Description:

My code works fine on jshell but I might be getting the wrong service here. 
Has anyone gotten this error too?

## Code Snippets

Attach code snippets (if any)

## Screenshots

<img width="1152" alt="Screenshot 2023-02-19 at 12 45 10 PM" src="https://user-images.githubusercontent.com/122283133/219924198-ebb6b69c-fc43-4dbd-8084-fbfa3a4dd694.png">

**Comment by leezhanpeng** [comment]
Hmm that's weird. Have you tried resubmitting?

**Comment by jaejayrome** [comment]
<img width="1256" alt="Screenshot 2023-02-18 at 2 07 16 PM" src="https://user-images.githubusercontent.com/80029422/219939734-98adb4f7-609e-4cdf-bf19-1ff85105eb04.png">

Same here Jeff, i can't pass this test case due to this as well!

**Comment by jaejayrome** [comment]
<img width="1426" alt="Screenshot 2023-02-19 at 5 38 17 PM" src="https://user-images.githubusercontent.com/80029422/219940336-c6891eb7-a5c0-4e6b-9f3d-daf935ba2e24.png">
Hi Jeff i resubmitted my code and received a compilation warning which im trying to fix now; perhaps it's due to this

**Comment by jeffkhong11** [comment]
Hi Jerome, I realised that my booking class was not general enough to extend to additional drivers and services. Changed it and it's working now.

**Comment by jaejayrome** [comment]
What do you mean by this? You mean that drivers and services should extend Booking?

**Comment by jeffkhong11** [comment]
> What do you mean by this? You mean that drivers and services should extend Booking?

I didn't do that. For me, I created an ImList<Service> for drivers to store their possible services, then loop through to create new bookings.

**Comment by jaejayrome** [comment]
thanks jeff, i realised where was my mistake already because my booking class and request classes weren't general enough like u said!

**Comment by xZealthiusx** [comment]
Hello Jeff, I am currently facing the same issue. Do you feel that the ImList of possible services should be stored in the Booking Class or Driver class? Or does it make no difference?

**Comment by jeffkhong11** [comment]
> Hello Jeff, I am currently facing the same issue. Do you feel that the ImList of possible services should be stored in the Booking Class or Driver class? Or does it make no difference?

Yes, you can do so in the Driver class. So u can call the method for that ImList in the constructor.

**Comment by nicholasteow** [comment]
wow, this was the exact issue I faced as well. I had similar ideas on the implementations and now I'm so glad I visited GitHub before implementing my classes so that I didn't have to go through all these as you did.

Thanks Jeff !!

**Comment by jeffkhong11** [comment]
> wow, this was the exact issue I faced as well. I had similar ideas on the implementations and now I'm so glad I visited GitHub before implementing my classes so that I didn't have to go through all these as you did.
> 
> Thanks Jeff !!

Glad to help Nicholas!

**Comment by jeffkhong11** [comment]
> > > What do you mean by this? You mean that drivers and services should extend Booking?
> > 
> > 
> > I didn't do that. For me, I created an ImList for drivers to store their possible services, then loop through to create new bookings.
> 
> I have the same issue here, but I still could not figure out what is wrong. In my driver class, I have a private final of immutable services which a particular driver can take. Then in my booking class, I loop through the immutable list of services to find the cheapest fare. Is that not general enough?

Hmm, is the problem in the constructor? For me, I did this.services = this.method();
then in the class itself: 
ImList<Service> method() { ... }

**Comment by jiarui22** [comment]
> > > What do you mean by this? You mean that drivers and services should extend Booking?
> > 
> > 
> > I didn't do that. For me, I created an ImList for drivers to store their possible services, then loop through to create new bookings.
> 
> I have the same issue here, but I still could not figure out what is wrong. In my driver class, I have a private final of immutable services which a particular driver can take. Then in my booking class, I loop through the immutable list of services to find the cheapest fare. Is that not general enough?

Is that because of private final?  What I did is create a method which returns the immutable list of the service that the driver can take.
